### ISSUE TITLE
Data table viewer control that returns value or index of selected item to ipython kernel

### ISSUE THREAD

**Issue by ianhelle** [microsoft/vscode-jupyter#5915]
We have a common scenario where we present a set of results in a data table view.
We'd then like a user to be able to select an item (either a row or a cell value) and be able to use this selection in a subsequent cell. There are several Jupyter extensions or Python packages that give you a nice filterable/sortable data table view, even selection of an item. E.g. the Bokeh DataTable. However, there is nothing I've found that will communicate selected values back to Python. 
IPyWidgets does this with its widgets.
In classic Jupyter you could create a hacky solution using a Javascript callback to have the IPython kernel execute a Python statement. This doesn't seem to work in VSCode native Jupyter - presumably because arbitrary JS execution is blocked.

I was wondering if this is something that you could create a Jupyter native widget for - perhaps based on the nteract data viewer.

**Comment by joyceerhl** [comment]
Hi @ianhelle, just to clarify for my understanding, sounds like the ask is for our team to create a table widget with the ability to register custom callbacks? Could you clarify what you would expect the callback interaction to look like?

**Comment by ianhelle** [comment]
Hi Joyce,
I spoke with @***@***.***> about this yesterday.

Essentially, we'd like a data table widget that behaves like an IPyWidget - in that it maintains synchronized state between the JS model and the Python object.

An interface a little like the Select widget (or SelectMultiple) but capable of displaying a table such as a pandas Df, would be a huge thing.

  *   value property returns the currently selected row, column or cell value (in the case of the row or column an index would prob be sufficient)
  *   observe method where you could register a callback to fire when the selection value changes
     *   I guess this could fire on other events like filter or sort changing
     *   datatable.observe(my_callback, names="value")  or datatable.observe(my_callback, names="selection") - would call my_callback with the state change dict with the new selection.

I realize this is not really a Vscode/jupyter core feature area.

An alternative would be a mechanism that would allow us to propogate JS state changes back to the IPython kernel. An example might be taking something like the Bokeh Datatable widget and have its jscallbacks set/change state in a custom ipywidget. My knowledge of JS and the JS model of notebooks and embedded widgets is pretty basic, so I haven't really experimented with this.

________________________________

**Comment by rchiodo** [comment]
This sounds similar to what nteract does with their data explorer. 

This blog post might be helpful
https://blog.nteract.io/designing-the-nteract-data-explorer-f4476d53f897

**Comment by ianhelle** [comment]
I may not have been clear in my previous description - unless I'm missing something in the blog post.

I *know* that there are several data table controls around, including the nteract one, qgrid, bokeh DataTable and others, that can display and filter data in a tabular format.

The key functionality that I was after is being able retrieve the value (e.g. a row) selected in the data viewer. All of the viewers that I've come across work in JS and have no mechanism to communicate state data back to the IPyKernel.
ipwidgets does this by implementing a MVC synchronization that keeps front and backends in sync.

This example might illustrate what I'm after.
```
In:  [1]  data_widget = DataViewer(data=my_df)
          data_widget
Out: [1] ...user browses the data table and selects row 7 (or cell with the contents "Arizona123")

In:  [2] data_widget.value
Out: [2] 7

or
In:  [2] data_widget.value
Out: [2] 'Arizona123'
```
If it followed the pattern of ipywidgets there would also be an observe function where you could register a Python callback on a "select" event (as the Select and SelectMultiple widgets do)

**Comment by rchiodo** [comment]
So I think what you're asking for is a control that updates some state in the kernel that you can then reference later in other cells?

In your example above that would be ```data_widget.value```?

And you're hoping the jupyter extension would provide this control? Or maybe integrate this into our existing control?

Which would likely mean that on selecting rows in the javascript item, we'd have to change some state in the kernel?

**Comment by ianhelle** [comment]
Hey sorry, I thought you'd consigned this issue to the trash can.
The issue came out of a discussion with Claudia. She suggested filing an issue here - although I agree that this doesn't feel like a python extension feature.
That is what I'm asking for - to allow JS controls to be able to propagate state change back to the kernel. It's kind of difficult to know what the right forum to ask this question would be. I was hoping that it might be possible to leverage the IPywidgets sync mechanism to have callbacks via traitlets. This sort of works (having Bokeh set a value of an IPywidget object) in Jupyter classic but not in other Jupyter environments like Jupyterlab, nteract or vscode.

**Comment by rchiodo** [comment]
It sounds like I'm misunderstanding then. 

You want _your_ control to be able to talk to a kernel?

I believe this is possible today. I think you could just use the WidgetManager (which should be a global object in the VS code DOM too) to call [_create_comm](https://github.com/jupyter-widgets/ipywidgets/blob/e0d41f6f02324596a282bc9e4650fd7ba63c0004/jupyterlab_widgets/src/manager.ts#L180). That would allow you to send messages to the kernel. 

The easiest way I think would be to write an ipywidget. Then all the comm infrastructure would just work. 

Although I'm still not sure if this is what you want.

Are you creating a new data table? Or do you want to just be able to listen to our data table in your notebook cells?

**Comment by rchiodo** [comment]
For me it's unclear what the ask is.

Do you want:
1) To provide your own dataviewer that can talk to the kernel. You'd have to implement the [observe](https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Events.html#Traitlet-events) functionality. I don't think this is what you're asking for, but just want to make sure. This could be done today if you wrote an ipywidget based viewer. You can even hook it into say the default viewer for pandas dataframes.
2) For the Jupyter extension to do all of this. Then you'd just be writing the observe code in your notebook.

**Comment by ianhelle** [comment]
> You want _your_ control to be able to talk to a kernel?
I think it probably is what I want - we have the Bokeh data table control, which works well.
I did try a bit of code to have a bokeh CustomJS event handler to relay the selected value to a custom widget but the WidgetManager didn't seem to be callable from the Bokeh JS code. It's quite likely that I got the code wrong though.
I'd started to assume that this wasn't possible without writing an extension.

I'll give this another go. 

On 2. - I have heard this request before, so, if there was some kind of "data transport" widget that's directly usable - albeit you'd have to do the JS callback plumbing it might be helpful to us Python-but-not-JS devs. Or maybe I should just spend more time learning JS.
